File: public/java/048-pacific-atlantic-water-flow.java
/*
 * LeetCode Problem #417: Pacific Atlantic Water Flow
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/pacific-atlantic-water-flow/
 */
import java.util.*;

public class Solution {

    private static final int[][] DIRS = {{0,1},{0,-1},{1,0},{-1,0}};

    // APPROACH 1: Brute Force DFS from Each Cell | O(m*n*(m+n)) time | O(m*n) space
    // EXPLAIN: For each cell independently DFS to check reachability to both oceans.
    public List<List<Integer>> pacificAtlanticBrute(int[][] h) {
        int m=h.length, n=h[0].length;
        List<List<Integer>> res = new ArrayList<>();
        for (int r=0;r<m;r++) for (int c=0;c<n;c++)
            if (canReach(r,c,h,m,n,true) && canReach(r,c,h,m,n,false))
                res.add(Arrays.asList(r,c));
        return res;
    }
    private boolean canReach(int r, int c, int[][] h, int m, int n, boolean pacific) {
        boolean[][] vis = new boolean[m][n];
        Deque<int[]> stack = new ArrayDeque<>();
        stack.push(new int[]{r,c}); vis[r][c]=true;
        while (!stack.isEmpty()) {
            int[] cell = stack.pop();
            int x=cell[0], y=cell[1];
            if (pacific  && (x==0||y==0))   return true;
            if (!pacific && (x==m-1||y==n-1)) return true;
            for (int[] d:DIRS) {
                int nx=x+d[0],ny=y+d[1];
                if (nx>=0&&nx<m&&ny>=0&&ny<n&&!vis[nx][ny]&&h[nx][ny]<=h[x][y]) {
                    vis[nx][ny]=true; stack.push(new int[]{nx,ny});
                }
            }
        }
        return false;
    }

    // APPROACH 2: Multi-Source BFS (Optimal) | O(m*n) time | O(m*n) space
    // EXPLAIN: BFS outward from each ocean's border cells (water flows uphill); intersect both reachability sets.
    public List<List<Integer>> pacificAtlantic(int[][] h) {
        int m=h.length, n=h[0].length;
        boolean[][] pac=new boolean[m][n], atl=new boolean[m][n];
        Queue<int[]> pq=new LinkedList<>(), aq=new LinkedList<>();
        for (int r=0;r<m;r++) { pac[r][0]=true; pq.offer(new int[]{r,0}); atl[r][n-1]=true; aq.offer(new int[]{r,n-1}); }
        for (int c=0;c<n;c++) { pac[0][c]=true; pq.offer(new int[]{0,c}); atl[m-1][c]=true; aq.offer(new int[]{m-1,c}); }
        bfs(pq,pac,h,m,n); bfs(aq,atl,h,m,n);
        List<List<Integer>> res = new ArrayList<>();
        for (int r=0;r<m;r++) for (int c=0;c<n;c++) if (pac[r][c]&&atl[r][c]) res.add(Arrays.asList(r,c));
        return res;
    }
    private void bfs(Queue<int[]> q, boolean[][] vis, int[][] h, int m, int n) {
        while (!q.isEmpty()) {
            int[] cell=q.poll(); int r=cell[0],c=cell[1];
            for (int[] d:DIRS) {
                int nr=r+d[0],nc=c+d[1];
                if (nr>=0&&nr<m&&nc>=0&&nc<n&&!vis[nr][nc]&&h[nr][nc]>=h[r][c]) { vis[nr][nc]=true; q.offer(new int[]{nr,nc}); }
            }
        }
    }

    // APPROACH 3: Multi-Source DFS (Optimal DFS variant) | O(m*n) time | O(m*n) space
    // EXPLAIN: Same reverse-flood idea using DFS instead of BFS.
    public List<List<Integer>> pacificAtlanticDFS(int[][] h) {
        int m=h.length, n=h[0].length;
        boolean[][] pac=new boolean[m][n], atl=new boolean[m][n];
        for (int r=0;r<m;r++) { dfsPAC(r,0,pac,h,m,n); dfsPAC(r,n-1,atl,h,m,n); }
        for (int c=0;c<n;c++) { dfsPAC(0,c,pac,h,m,n); dfsPAC(m-1,c,atl,h,m,n); }
        List<List<Integer>> res = new ArrayList<>();
        for (int r=0;r<m;r++) for (int c=0;c<n;c++) if (pac[r][c]&&atl[r][c]) res.add(Arrays.asList(r,c));
        return res;
    }
    private void dfsPAC(int r, int c, boolean[][] vis, int[][] h, int m, int n) {
        vis[r][c]=true;
        for (int[] d:DIRS) {
            int nr=r+d[0],nc=c+d[1];
            if (nr>=0&&nr<m&&nc>=0&&nc<n&&!vis[nr][nc]&&h[nr][nc]>=h[r][c]) dfsPAC(nr,nc,vis,h,m,n);
        }
    }
}

// Made with Bob

File: public/python/048-pacific-atlantic-water-flow.py
# LeetCode Problem #417: Pacific Atlantic Water Flow
# Difficulty: Medium
# Link: https://leetcode.com/problems/pacific-atlantic-water-flow/

from typing import List
from collections import deque


# ─────────────────────────────────────────────
# APPROACH 1: Brute Force DFS from Each Cell | O(m*n*(m+n)) time | O(m*n) space
# EXPLAIN: For each cell independently check whether it can reach both oceans via DFS.
# WHEN: Conceptually clearest; too slow for large grids but good for understanding.

def pacificAtlantic_brute(heights: List[List[int]]) -> List[List[int]]:
    m, n = len(heights), len(heights[0])
    dirs = [(0,1),(0,-1),(1,0),(-1,0)]

    def can_reach(r, c, pacific):
        visited = [[False]*n for _ in range(m)]
        stack = [(r, c)]
        visited[r][c] = True
        while stack:
            x, y = stack.pop()
            if pacific and (x == 0 or y == 0):
                return True
            if not pacific and (x == m-1 or y == n-1):
                return True
            for dx, dy in dirs:
                nx, ny = x+dx, y+dy
                if 0 <= nx < m and 0 <= ny < n and not visited[nx][ny] and heights[nx][ny] <= heights[x][y]:
                    visited[nx][ny] = True
                    stack.append((nx, ny))
        return False

    return [[r, c] for r in range(m) for c in range(n)
            if can_reach(r, c, True) and can_reach(r, c, False)]


# ─────────────────────────────────────────────
# APPROACH 2: Multi-Source BFS (Optimal) | O(m*n) time | O(m*n) space
# EXPLAIN: Reverse-flood from each ocean border (water flows uphill from ocean inward). Cells reachable from both = answer.
# WHEN: Optimal; single BFS per ocean instead of per cell.

def pacificAtlantic_bfs(heights: List[List[int]]) -> List[List[int]]:
    m, n = len(heights), len(heights[0])
    dirs = [(0,1),(0,-1),(1,0),(-1,0)]

    def bfs(starts):
        visited = [[False]*n for _ in range(m)]
        q = deque(starts)
        for r, c in starts:
            visited[r][c] = True
        while q:
            r, c = q.popleft()
            for dr, dc in dirs:
                nr, nc = r+dr, c+dc
                if 0 <= nr < m and 0 <= nc < n and not visited[nr][nc] and heights[nr][nc] >= heights[r][c]:
                    visited[nr][nc] = True
                    q.append((nr, nc))
        return visited

    pac_starts = [(0, c) for c in range(n)] + [(r, 0) for r in range(1, m)]
    atl_starts = [(m-1, c) for c in range(n)] + [(r, n-1) for r in range(m-1)]
    pac = bfs(pac_starts)
    atl = bfs(atl_starts)
    return [[r, c] for r in range(m) for c in range(n) if pac[r][c] and atl[r][c]]


# ─────────────────────────────────────────────
# APPROACH 3: Multi-Source DFS (Optimal DFS variant) | O(m*n) time | O(m*n) space
# EXPLAIN: Same reverse-flood idea using DFS from border cells.
# WHEN: Same time complexity as BFS; DFS uses the call stack instead of an explicit queue.

def pacificAtlantic_dfs(heights: List[List[int]]) -> List[List[int]]:
    m, n = len(heights), len(heights[0])
    dirs = [(0,1),(0,-1),(1,0),(-1,0)]

    def dfs(r, c, visited):
        visited[r][c] = True
        for dr, dc in dirs:
            nr, nc = r+dr, c+dc
            if 0 <= nr < m and 0 <= nc < n and not visited[nr][nc] and heights[nr][nc] >= heights[r][c]:
                dfs(nr, nc, visited)

    pac = [[False]*n for _ in range(m)]
    atl = [[False]*n for _ in range(m)]
    for c in range(n):
        dfs(0, c, pac)
        dfs(m-1, c, atl)
    for r in range(m):
        dfs(r, 0, pac)
        dfs(r, n-1, atl)
    return [[r, c] for r in range(m) for c in range(n) if pac[r][c] and atl[r][c]]


# Made with Bob
